### PR TITLE
feat(stonedb): support exists/in subquery convert to semi-join (#864)

### LINKED ISSUE #864: feature:  support semi-join in stonedb, include EXISTS and IN subqueries
**Is your feature request related to a problem? Please describe.**

The reason why our tpch Q4 query is slow is because the execution strategy of subquery is only one kind of nested loop, here we want to do an optimization is to convert the subquery expansion to semi-join, can use hash join, so that we can use hash-join to speed up.
**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by duanfuxiang0**
## Summary about this PR

Issue Number: close #864 

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [x] New Feature
- [ ] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: mysql-test/suite/tianmu/r/select_expressions.result [hunk 1]
```diff
@@ -53,8 +53,6 @@ t1_int	t1_char
 SELECT * FROM t1 WHERE t1_char IN (SELECT t1_char FROM t1 WHERE t1_char='aaa');
 t1_int	t1_char
 1	aaa
-1	aaa
-2	aaa
 2	aaa
 SELECT CASE t1_int WHEN t1_int=1 THEN t1_char ELSE t1_int END FROM t1;
```
Comment (RingsC) [inline mysql-test/suite/tianmu/r/select_expressions.result:57]: such these test cases should be added .
```
1. IN/=ANY predicates on the form:

  @code
  SELECT ...
  FROM ot1 ... otN
  WHERE (oe1, ... oeM) IN (SELECT ie1, ..., ieM
                           FROM it1 ... itK
                          [WHERE inner-cond])
   [AND outer-cond]
  [GROUP BY ...] [HAVING ...] [ORDER BY ...]
  @endcode

  are transformed into:

  @code
  SELECT ...
  FROM (ot1 ... otN) SJ (it1 ... itK)
                     ON (oe1, ... oeM) = (ie1, ..., ieM)
                        [AND inner-cond]
  [WHERE outer-cond]
  [GROUP BY ...] [HAVING ...] [ORDER BY ...]
  @endcode

  Notice that the inner-cond may contain correlated and non-correlated
  expressions. Further transformations will analyze and break up such
  expressions.

  2. EXISTS predicates on the form:

  @code
  SELECT ...
  FROM ot1 ... otN
  WHERE EXISTS (SELECT expressions
                FROM it1 ... itK
                [WHERE inner-cond])
   [AND outer-cond]
  [GROUP BY ...] [HAVING ...] [ORDER BY ...]
  @endcode

  are transformed into:

  @code
  SELECT ...
  FROM (ot1 ... otN) SJ (it1 ... itK)
                     [ON inner-cond]
  [WHERE outer-cond]
  [GROUP BY ...] [HAVING ...] [ORDER BY ...]
  @endcode

  3. Negated EXISTS predicates on the form:

  @code
  SELECT ...
  FROM ot1 ... otN
  WHERE NOT EXISTS (SELECT expressions
                FROM it1 ... itK
                [WHERE inner-cond])
   [AND outer-cond]
  [GROUP BY ...] [HAVING ...] [ORDER BY ...]
  @endcode
```

File: sql/sql_resolver.cc [hunk 2]
```diff
@@ -1022,6 +1032,19 @@ bool SELECT_LEX::resolve_subquery(THD *thd)
     }
   }
 
+  bool tianmu_table_has_index = false;
+  SQL_I_List<TABLE_LIST> parent_tables = parent_lex->select_lex->table_list;
+  for (TABLE_LIST*p_table_list = parent_tables.first; p_table_list;
```
Comment (RingsC) [inline sql/sql_resolver.cc]: Add a space between `TABLE_LIST*` and `p_table_list`

Comment (RingsC) [reply sql/sql_resolver.cc]: Do we just to set the flag to true ONLY via one tianmu table with PK?

**Review by RingsC — CHANGES_REQUESTED** [review]
Pls, adds the test cases firstly to make sure we have cover all corner cases.

File: sql/sql_resolver.cc (hunk 2 above)
Comment (duanfuxiang0) [reply sql/sql_resolver.cc]: No need, it will automatically determine whether the tianmu's table contains a primary key

File: mysql-test/suite/tianmu/r/select_expressions.result (hunk 1 above)
Comment (duanfuxiang0) [reply mysql-test/suite/tianmu/r/select_expressions.result:57]: ok

**Review by RingsC — APPROVED** [review]

**Review by isredstar — APPROVED** [review]
LGTM
